src/Channels.Networking.Sockets/Signal.cs

```diff
@@ -0,0 +1,173 @@
+﻿using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace Channels.Networking.Sockets
+{
+    internal enum ContinuationMode
+    {
+        Synchronous,
+        ThreadPool,
+        // TODO: sync-context? but if so: whose? the .Current at creation? at SetResult?
+    }
+    internal class Signal : INotifyCompletion
+    {
+
+        internal Signal WaitAsync() => this;
+        public Signal GetAwaiter() => this; // we can ourselves be an awaiter, why not?
+        private Action _continuation; // interlocked access
+        private readonly ContinuationMode _continuationMode;
+        private volatile bool _isCompleted;
+
+        public Signal(ContinuationMode continuationMode = ContinuationMode.Synchronous)
```

Comment:
Phew, David would be insufferable if he knew that I just found and fixed a critical major bug in this code as a result of this advice. Fortunately, he'll never know...